File: src/table.rs
use crate::constants::*;
use core::num;
use std::io::{Read, Seek, SeekFrom};
use std::{
    fs::{File, OpenOptions},
    io::Write,
};

#[derive(Debug)]
pub struct Row {
    pub id: u32,
    pub name: [u8; 32],
    pub email: [u8; 255],
}

impl Row {
    pub fn new() -> Row {
        Row {
            id: 0,
            name: [0; USERNAME_SIZE],
            email: [0; EMAIL_SIZE],
        }
    }

    pub fn show(&self) {
        println!(
            "({},{},{})",
            self.id,
            std::str::from_utf8(&self.name)
                .unwrap()
                .trim_end_matches(char::from(0)),
            std::str::from_utf8(&self.email)
                .unwrap()
                .trim_end_matches(char::from(0))
        );
    }
}

struct Pager {
    file: File,          // DB文件
    file_length: u64,    // 文件长度
    pages: Vec<Vec<u8>>, // 数据页
}

impl Pager {
    fn new() -> Pager {
        Pager {
            file: File::create("xx").unwrap(),
            file_length: 0,
            pages: vec![vec![]; TABLE_MAX_PAGES],
        }
    }

    fn new_from_file(filename: &str) -> Pager {
        match OpenOptions::new()
            .read(true)
            .write(true)
            .create(true)
            .open(filename)
        {
            Ok(mut file) => {
                let file_length = file.seek(SeekFrom::End(0)).unwrap();
                println!("file length is {}", file_length);
                Pager {
                    file,
                    file_length,
                    pages: vec![vec![]; TABLE_MAX_PAGES],
                }
            }
            Err(e) => {
                println!("open file err {}", e);
                std::process::exit(1);
            }
        }
    }

    fn get_page(&mut self, page_num: u32) {
        if page_num > TABLE_MAX_PAGES as u32 {
            println!(
                "page number out of bounds. {} > {}",
                page_num, TABLE_MAX_PAGES
            );
            std::process::exit(1);
        }
        if self.pages[page_num as usize].is_empty() {
            // println!("page-> {} empty!", page_num);

            let mut nums_pages = self.file_length / PAGE_SIZE as u64;

            if self.file_length % PAGE_SIZE as u64 > 0 {
                nums_pages += 1;
            }

            if page_num as u64 <= nums_pages {
                if self
                    .file
                    .seek(SeekFrom::Start((page_num * PAGE_SIZE as u32) as u64))
                    .is_err()
                {
                    println!("seeking file err");
                    std::process::exit(1);
                }
                let buf_size = if self.file_length >= (page_num * PAGE_SIZE as u32) as u64 {
                    (self.file_length - (page_num * PAGE_SIZE as u32) as u64) as usize
                } else {
                    // PAGE_SIZE
                    // self.file_length as usize
                    PAGE_SIZE
                };
                let mut page = vec![0u8; buf_size];

                if self.file.read_exact(page.as_mut_slice()).is_err() {
                    println!("error reading file {}", page.len());
                    std::process::exit(1);
                }
                // TODO
                // self.pages[page_num as usize].extend(page.as_slice());
                self.pages[page_num as usize].extend_from_slice(page.as_slice());
            }
        } else {
            // println!("page-> {} alloced!", page_num);
        }
    }

    fn flush(&mut self, page_num: u32) {
        if self.pages[page_num as usize].is_empty() {
            println!("Tried to flush null page");
            std::process::exit(1);
        }

        if self
            .file
            .seek(SeekFrom::Start((page_num * PAGE_SIZE as u32) as u64))
            .is_err()
        {
            println!("Error seeking.");
            std::process::exit(1);
        }

        let drained_vec: Vec<u8> = self.pages[page_num as usize].drain(..).collect();
        self.pages[page_num as usize].shrink_to_fit();

        if self.file.write_all(drained_vec.as_ref()).is_err() {
            println!("Error writing.");
            std::process::exit(1);
        }
    }
}

pub struct Table {
    pub rows_count: u32,
    // pages: Vec<Vec<u8>>,
    pager: Pager,
}

impl Table {
    pub fn new() -> Table {
        Table {
            rows_count: 0,
            // pages: vec![vec![]; TABLE_MAX_PAGES as usize],
            pager: Pager::new(),
        }
    }

    pub fn db_open(filename: &str) -> Table {
        let pager = Pager::new_from_file(filename);
        let rows_count = (pager.file_length / ROW_SIZE as u64) as u32;
        Table { rows_count, pager }
    }

    pub fn row_slot(&mut self, row_count: u32) -> (u32, u32) {
        let page_num = row_count / ROWS_PER_PAGE as u32;
        let row_offset = row_count % ROWS_PER_PAGE as u32;
        let byte_offset = row_offset * ROW_SIZE as u32;
        // println!(
        //     "slot: page_num={}, row_offset={}, byte_offset={}",
        //     page_num, row_offset, byte_offset
        // );
        self.pager.get_page(page_num);
        return (page_num, byte_offset);
    }

    pub fn serialize_row(&mut self, row: &Row, page_num: u32, byte_offsets: u32) {
        let id_bytes = row.id.to_ne_bytes();
        let name_bytes = row.name;
        let email_bytes = row.email;

        // println!("{:?}, {:?}, {:?}", &id_bytes, &name_bytes, &email_bytes);

        let offset = byte_offsets as usize;

        // self.pager.pages[page_num as usize][(offset + ID_OFFSET)..(offset + ID_OFFSET + ID_SIZE)]
        //     .copy_from_slice(&id_bytes);
        // self.pager.pages[page_num as usize]
        //     [(offset + USERNAME_OFFSET)..(offset + USERNAME_OFFSET + USERNAME_SIZE)]
        //     .copy_from_slice(&name_bytes);
        // self.pager.pages[page_num as usize]
        //     [(offset + EMAIL_OFFSET)..(offset + EMAIL_OFFSET + EMAIL_SIZE)]
        //     .copy_from_slice(&email_bytes);

        self.pager.pages[page_num as usize].extend_from_slice(&id_bytes);
        self.pager.pages[page_num as usize].extend_from_slice(&name_bytes);
        self.pager.pages[page_num as usize].extend_from_slice(&email_bytes);

        // println!("{:?}", &self.pager.pages[page_num as usize]);
    }

    pub fn deserialize_row(&mut self, page_num: u32, byte_offsets: u32) -> Row {
        let offset = byte_offsets as usize;
        let id_bytes_slice = &self.pager.pages[page_num as usize]
            [(offset + ID_OFFSET)..(offset + ID_OFFSET + ID_SIZE)];
        let name_bytes_slice = &self.pager.pages[page_num as usize]
            [(offset + USERNAME_OFFSET)..(offset + USERNAME_OFFSET + USERNAME_SIZE)];
        let email_bytes_slice = &self.pager.pages[page_num as usize]
            [(offset + EMAIL_OFFSET)..(offset + EMAIL_OFFSET + EMAIL_SIZE)];

        let mut id_bytes_arr = [0; 4];
        id_bytes_arr.copy_from_slice(id_bytes_slice);
        let id = u32::from_ne_bytes(id_bytes_arr);
        let mut name = [0u8; USERNAME_SIZE];
        name.copy_from_slice(name_bytes_slice);
        let mut email = [0u8; EMAIL_SIZE];
        email.copy_from_slice(email_bytes_slice);

        Row { id, name, email }
    }

    pub fn db_close(&mut self) {
        let num_full_pages = self.rows_count / ROWS_PER_PAGE as u32;

        for i in 0..num_full_pages {
            if self.pager.pages[i as usize].is_empty() {
                continue;
            }
            self.pager.flush(i);
        }

        let num_add_rows = self.rows_count % ROWS_PER_PAGE as u32;
        if num_add_rows > 0 {
            let page_num = num_full_pages;
            if !self.pager.pages[page_num as usize].is_empty() {
                self.pager.flush(page_num);
            }
        }

        if self.pager.file.sync_data().is_err() {
            println!("error closing db file.");
            std::process::exit(1);
        }
    }
}

File: c-impl/table.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <fcntl.h>  // open O_RDWR
// #include <sys/types.h>
#include <unistd.h> // lseek
#include <sys/stat.h>  // S_IWUSR
#include <errno.h>
#include "table.h"
#include "btree.h"

// #define size_of_attribute(Struct, Attribute) sizeof(((Struct*)0) -> Attribute)
// const uint32_t ID_SIZE = size_of_attribute(Row, id);
// const uint32_t USERNAME_SIZE = size_of_attribute(Row, username);
// const uint32_t EMAIL_SIZE = size_of_attribute(Row, email);

// const uint32_t ID_SIZE = sizeof((Row){.id=0, .username={}, .email={}}.id);
// const uint32_t USERNAME_SIZE = sizeof((Row){0, {}, {}}.username);
// const uint32_t EMAIL_SIZE = sizeof((Row){0, {}, {}}.email);

// 将 Row数据 序列化
void serialize_row(Row* row, void* dst) {
    // 从src 复制 N 个字节 到 dst
    memcpy(dst+ID_OFFSET, &(row->id), ID_SIZE);
    memcpy(dst+USERNAME_OFFSET, &(row->username), USERNAME_SIZE);
    memcpy(dst+EMAIL_OFFSET, &(row->email), EMAIL_SIZE);
}

// 将目标地址的数据，反序列化到 Row 结构
void deserialize_row(void* src, Row* row) {
    memcpy(&(row->id), src+ID_OFFSET, ID_SIZE);
    memcpy(&(row->username), src+USERNAME_OFFSET, USERNAME_SIZE);
    memcpy(&(row->email), src+EMAIL_OFFSET, EMAIL_SIZE);
}

// 创建新表
// Table* new_table() {
//     Table* table = malloc(sizeof(Table));
//     table->num_rows = 0;
//     for (uint32_t i = 0; i < TABLE_MAX_PAGES; i++) {
//         table->pages[i] = NULL;
//     }
//     return table;
// }

// 释放表占用的内存
// void free_table(Table* table) {
//     for (uint32_t i = 0; i < TABLE_MAX_PAGES; i++) {
//         free(table->pages[i]);
//     }
//     free(table);
// }

// 打印行记录
void print_row(Row* row) {
    printf("(%d, %s, %s)\n", row->id, row->username, row->email);
}

// 打开DB文件 构建Pager
Pager* pager_open(const char* filename) {
    int fd = open(filename,
                    O_RDWR |  // 读写模式
                        O_CREAT, // 创建文件如果文件不存在
                    S_IWUSR | // 用户写权限
                        S_IRUSR); // 用户读权限
    if (fd == -1) {
        printf("Unable to open file\n");
        exit(EXIT_FAILURE);
    }
    // 文件长度
    off_t file_length = lseek(fd, 0, SEEK_END);

    Pager* pager = malloc(sizeof(Pager));
    pager->fd = fd;
    pager->file_length = file_length;
    pager->num_pages = (file_length / PAGE_SIZE);

    // Page使用Btree存储后，文件长度必定是PAGE_SIZE的整数倍
    if (file_length % PAGE_SIZE != 0) {
        printf("Db file is not a whole number of pages. Corrupt file.\n");
        exit(EXIT_FAILURE);
    }

    for (uint32_t i = 0; i < TABLE_MAX_PAGES; i++) {
        pager->pages[i] = NULL;
    }

    return pager;
}

// 初始化 pager 结构， 初始化 table 结构
Table* db_open(const char* filename) {
    Pager* pager = pager_open(filename);
    // 文件长度/每一条行记录的大小 = 行记录的总数
    // uint32_t num_rows = pager->file_length /  ROW_SIZE;

    Table* table = malloc(sizeof(Table));
    table->pager = pager;
    // table->num_rows = num_rows;
    table->root_page_num = 0;

    if (pager->num_pages == 0) {
        // 初始化 Page0 作为第一个叶节点
        void* node = get_page(pager, 0);
        initialize_leaf_node(node);
    }

    return table;
}

// 页缓存未命中时，申请内存
// 当flush时，讲数据刷到磁盘文件中
void* get_page(Pager* pager, uint32_t page_num) {
    if (page_num > TABLE_MAX_PAGES) {
        printf("Tried to fetch page number out of bounds."
            "%d > %d\n", page_num, TABLE_MAX_PAGES);
        exit(EXIT_FAILURE);
    }
    // cache missing
    // 申请内存，并且从文件中加载数据
    if (pager->pages[page_num] == NULL) {
        void* page = malloc(PAGE_SIZE);
        // 文件长度 / 每页大小 = 页数[下阈]
        uint32_t num_pages = pager->file_length / PAGE_SIZE;
        // 所以也可能需要保存文件最后的部分页
        if (pager->file_length % PAGE_SIZE != 0) {
            num_pages += 1;
        }

        if (page_num <= num_pages) {
            // 调整文件当前指针到所需页的开头 [页数从0开始]
            lseek(pager->fd, page_num * PAGE_SIZE, SEEK_SET);
            // 读取一整页的内容到 page [刚刚申请好的内存]
            ssize_t bytes_read = read(pager->fd, page, PAGE_SIZE);
            if (bytes_read == -1) {
                printf("Error reading file: %d\n", errno);
                exit(EXIT_FAILURE);
            }
        }

        pager->pages[page_num] = page;

        // 当前页数>= Pager的page数量, 调整page数量
        if (page_num >= pager->num_pages) {
            pager->num_pages = page_num + 1;
        }
    }

    return pager->pages[page_num];
}

// 将页内容刷到磁盘
// // 目前的设计文件长度和行记录条数相关，因此需要计算文件最后的部分页
// 改用btree设计page. 一个节点占用一页[即使一页未满]，因此不存在额外的部分页
// void pager_flush(Pager* pager, uint32_t page_num, uint32_t size) {
void pager_flush(Pager* pager, uint32_t page_num) {
    if (pager->pages[page_num] == NULL) {
        printf("Tried to flush null page\n");
        exit(EXIT_FAILURE);
    }

    // lseek 定位到文件中当前页偏移位置
    off_t offset = lseek(pager->fd, page_num * PAGE_SIZE, SEEK_SET);
    if (offset == -1) {
        printf("Error seeking: %d\n", errno);
        exit(EXIT_FAILURE);
    }

    // 写入一整个页的大小
    ssize_t bytes_written = write(pager->fd, pager->pages[page_num], PAGE_SIZE);

    if (bytes_written == -1) {
        printf("Error writing: %d\n", errno);
        exit(EXIT_FAILURE);
    }
}

// 关闭数据库
// 1. 将页缓存中的内容刷磁盘
// 2. 关闭DB文件
// 3. 释放内存[Pager && Table]
void db_close(Table* table) {
    Pager* pager = table->pager;

    // 改用btree, 不再有额外的部分页需要处理
    // // 表中总的记录数 / 一页的记录数 = 表中总页数[下阈]
    // uint32_t num_full_pages = table->num_rows / ROWS_PER_PAGE;

    for (uint32_t i = 0; i < pager->num_pages; i++) {
        if (pager->pages[i] == NULL) {
            // 空page
            continue;
        }
        // pager_flush(pager, i, PAGE_SIZE);
        pager_flush(pager, i);
        free(pager->pages[i]);
        pager->pages[i] = NULL;
    }
    
    // // 同样，可能余有部分页记录需要写到文件最后
    // uint32_t num_additional_rows = table->num_rows % ROWS_PER_PAGE;
    // if (num_additional_rows > 0) {
    //     uint32_t page_num = num_full_pages;
    //     if (pager->pages[page_num] != NULL) {
    //         pager_flush(pager, page_num, num_additional_rows * ROW_SIZE);
    //         free(pager->pages[page_num]);
    //         pager->pages[page_num] = NULL;
    //     }
    // }

    int ret_code = close(pager->fd);
    if (ret_code == -1) {
        printf("Error cloing db file.\n");
        exit(EXIT_FAILURE);
    }
    for (uint32_t i = 0; i < TABLE_MAX_PAGES; i++) {
        void* page = pager->pages[i];
        if (page != NULL) {
            free(page);
            pager->pages[i] = NULL;
        }
    }
    free(pager);
    free(table);
}
